core/drivers/crypto/caam/utils/utils_dmaobj.c

```diff
@@ -0,0 +1,1395 @@
+
+	entry = TAILQ_FIRST(&priv->list);
+	while (entry) {
+		DMAOBJ_TRACE("Is type 0x%" PRIx8 " newbuf %s", priv->type,
+			     entry->newbuf ? "true" : "false");
+
+		next = TAILQ_NEXT(entry, next);
+
+		DMAOBJ_TRACE("Free entry %p", entry);
+		caam_free(entry);
+
+		entry = next;
+		DMAOBJ_TRACE("Next entry %p", entry);
+	};
+
+	if (priv->nb_sgtbuf) {
+		DMAOBJ_TRACE("Free #%d SGT data %p", priv->nb_sgtbuf,
+			     priv->sgtdata);
+		caam_free(priv->sgtdata);
+
+		obj->sgtbuf.number = priv->nb_sgtbuf;
+		obj->sgtbuf.sgt_type = (priv->nb_sgtbuf > 1) ? true : false;
+	}
+
+	if (priv->dmabuf.allocated) {
+		DMAOBJ_TRACE("Free CAAM DMA buffer");
+		caam_free_buf(&priv->dmabuf.buf);
+	}
+
+	if (priv->type & DMAOBJ_ALLOC_ORIG) {
+		DMAOBJ_TRACE("Free Allocated origin");
+		caam_free_buf(&obj->orig);
+	}
+
+	DMAOBJ_TRACE("Free private object %p", priv);
+	caam_free(priv);
+
+end:
+	if (obj->sgtbuf.number) {
+		DMAOBJ_TRACE("Free #%d SGT/Buffer %p", obj->sgtbuf.number,
+			     &obj->sgtbuf);
+		caam_sgtbuf_free(&obj->sgtbuf);
+	}
+
+	memset(obj, 0, sizeof(*obj));
+
+	cpu_spin_unlock_xrestore(&memlock, exceptions);
+}
+
+TEE_Result caam_dmaobj_add_first_block(struct caamdmaobj *obj,
+				       struct caamblock *block)
+{
+	struct priv_dmaobj *priv = NULL;
+	struct caambuf newbuf = {};
+	struct dmaentry *entry = NULL;
+
+	if (!obj || !obj->priv || !block)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	priv = obj->priv;
+
+	/* Save the block buffer reference and insert it at the head list */
+	newbuf.data = block->buf.data;
+	newbuf.length = block->filled;
+	newbuf.paddr = block->buf.paddr;
+	newbuf.nocache = block->buf.nocache;
+
+	entry = dmalist_add_entry_head(priv, &newbuf);
+
+	if (!entry)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	/*
+	 * Block buffer added in the output DMA buffer doesn't have to
+	 * be part of the output copy to origin buffer.
+	 */
+	if (priv->type & DMAOBJ_OUTPUT)
+		entry->nocopy = true;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result caam_dmaobj_derive_sgtbuf(struct caamdmaobj *obj,
+				     const struct caamdmaobj *from,
+				     size_t offset, size_t length)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	enum caam_status retstatus = CAAM_FAILURE;
+	struct priv_dmaobj *priv = NULL;
+
+	DMAOBJ_TRACE("Derive object %p - offset %zu - length %zu bytes", from,
+		     offset, length);
+
+	if (!obj || !from || !length || !from->priv) {
+		ret = TEE_ERROR_BAD_PARAMETERS;
+		goto end;
+	}
+
+	if (!from->orig.data || !from->orig.length) {
+		DMAOBJ_TRACE("No data/length to derive from");
+		ret = TEE_ERROR_NO_DATA;
+		goto end;
+	}
+
+	priv = from->priv;
+	if (!priv->nb_sgtbuf) {
+		DMAOBJ_TRACE("From SGT/Buffer not prepared");
+		ret = TEE_ERROR_NO_DATA;
+		goto end;
+	}
+
+	retstatus =
+		caam_sgt_derive(&obj->sgtbuf, &from->sgtbuf, offset, length);
+
+	ret = caam_status_to_tee_result(retstatus);
+
+end:
+	DMAOBJ_TRACE("Object returns 0x%" PRIx32, ret);
+	return ret;
+}
+
+TEE_Result caam_dmaobj_prepare(struct caamdmaobj *input,
```

Comment:
This function has a repeated symmetry which make quite long a complicated.
Perhaps it would be possible to refactor with a couple of helper functions?

Reply:
Ok